src/common/state/task_table.c

```diff
@@ -30,6 +30,21 @@ void task_table_update(db_handle *db_handle,
                       done_callback, redis_task_table_update, user_context);
 }
 
+void task_table_test_and_update(db_handle *db_handle,
+                                OWNER task *task,
+                                scheduling_state test_state,
+                                retry_info *retry,
+                                task_table_get_callback done_callback,
+                                void *user_context) {
+  task_table_test_and_update_data *update_data =
+      malloc(sizeof(task_table_test_and_update_data));
+  update_data->task = task;
```

Comment:
Yeah, this is a little unfortunate...I had to do it this way because of the way callback data is cleaned up (the failure callback calls `free(callback_data->data)`). Let me see how it would look if we kept the entire task in `task_table_test_and_update_data` instead of a pointer.